File: Trees/Python/Maximum Number of K-Divisible Components.py
# Problem link: https://leetcode.com/problems/maximum-number-of-k-divisible-components/description
# Time: O(n + m)
# Space: O(n + m)
class Solution:
    def maxKDivisibleComponents(self, n: int, edges: List[List[int]], values: List[int], k: int) -> int:
        adjList, result = [[] for _ in range(n)], 0

        for u, v in edges:
            adjList[u].append(v)
            adjList[v].append(u)
        

        def dfs(par, node):
            nonlocal result
            total = values[node]

            for nei in adjList[node]:
                if nei != par:
                    total += dfs(node, nei)
            
            if total % k == 0:
                result += 1
                return 0
            
            return total

        dfs(-1, 0)

        return result

File: Trees/Java/Maximum Number of K-Divisible Components.java
// Problem link: https://leetcode.com/problems/maximum-number-of-k-divisible-components/description
// Time: O(n + m)
// Space: O(n + m)
class Solution {
    public int maxKDivisibleComponents(int n, int[][] edges, int[] values, int k) {
        List<Integer>[] adjList = new ArrayList[n];
        int[] result = new int[1];

        for(int index = 0; index < n; index++) {
            adjList[index] = new ArrayList<>();
        }

        for(int[] edge : edges) {
            adjList[edge[0]].add(edge[1]);
            adjList[edge[1]].add(edge[0]);
        }

        this.dfs(-1, 0, adjList, result, values, k);

        return result[0];
    }

    private int dfs(int par, int node, List<Integer>[] adjList, int[] result, int[] values, int k) {
        int total = values[node];

        for(int nei : adjList[node]) {
            if(nei != par) {
                total += this.dfs(node, nei, adjList, result, values, k) % k;
            }
        }

        if(total % k == 0) {
            result[0]++;
        }

        return total;
    }
}
